src/librustc/infer/error_reporting/nice_region_error/static_impl_trait.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright 2012-2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Error Reporting for static impl Traits.
+
+use infer::error_reporting::nice_region_error::NiceRegionError;
+use infer::lexical_region_resolve::RegionResolutionError;
+use ty::RegionKind;
+use util::common::ErrorReported;
+
+impl<'a, 'gcx, 'tcx> NiceRegionError<'a, 'gcx, 'tcx> {
+    /// Print the error message for lifetime errors when the return type is a static impl Trait.
+    pub(super) fn try_report_static_impl_trait(&self) -> Option<ErrorReported> {
+        if let Some(ref error) = self.error {
+            match error.clone() {
+                RegionResolutionError::SubSupConflict(
+                    var_origin,
+                    sub_origin,
+                    sub_r,
+                    sup_origin,
+                    sup_r,
+                ) => {
+                    let anon_reg_sup = self.is_suitable_region(sup_r)?;
+                    if sub_r == &RegionKind::ReStatic &&
```

Comment:
Are we just concluding that *any* time that we have a bound of `'static` and a fn with an impl Trait return, the two are connected?

e.g., what happens with this example?

```rust
use std::fmt::Debug;
fn foo(x: &u32) -> impl Debug {
  let _: &'static u32 = x;
  ()
}
```

Reply:
> Are we just concluding that any time that we have a bound of 'static and a fn with an impl Trait return, the two are connected?

Close, unless I misunderstand, the code checks that the return type is `impl Trait` _and_ that it is bound to `'static`.

> what happens with

The two provided examples keep the current output:

```
error[E0621]: explicit lifetime required in the type of `x`
 --> file.rs:3:25
  |
2 | fn foo(x: &u32) -> impl Debug {
  |        - consider changing the type of `x` to `&'static u32`
3 |   let _: &'static u32 = x;
  |                         ^ lifetime `'static` required

error[E0621]: explicit lifetime required in the type of `x`
 --> file.rs:7:29
  |
7 | fn bar(x: &u32, y: &u32) -> impl Debug {
  |        -                    ^^^^^^^^^^ lifetime `'static` required
  |        |
  |        consider changing the type of `x` to `&'static u32`
```

For the second case, we could expand this code handle it and suggest the following, but it'll be very involved:

```
fn bar<'a>(x: &'a u32, y: &u32) -> impl Debug + 'a {
  x
}
```